routing/index_graph.hpp

```diff
@@ -25,36 +33,159 @@ class JointEdge final
+
+    if (m_center != rhs.m_center)
+      return m_center < rhs.m_center;
+
+    if (m_to != rhs.m_to)
+      return m_to < rhs.m_to;
+
+    if (m_fromFeatureId != rhs.m_fromFeatureId)
+      return m_fromFeatureId < rhs.m_fromFeatureId;
+
+    return m_toFeatureId < rhs.m_toFeatureId;
+  }
+
+  bool operator==(RestrictionInfo const & rhs) const { return !(*this < rhs || rhs < *this); }
+
+  bool operator!=(RestrictionInfo const & rhs) const { return !(*this == rhs); }
+
+  pair<DirectedEdge, DirectedEdge> Edges() const
+  {
+    return make_pair(DirectedEdge(m_from, m_center, m_fromFeatureId),
+                     DirectedEdge(m_center, m_to, m_toFeatureId));
+  }
+
+  Joint::Id m_from = Joint::kInvalidId;
+  Joint::Id m_center = Joint::kInvalidId;
+  Joint::Id m_to = Joint::kInvalidId;
+  uint32_t m_fromFeatureId = 0;
+  uint32_t m_toFeatureId = 0;
 };
 
 class IndexGraph final
 {
 public:
+  static uint32_t const kStartFakeFeatureIds = 1024 * 1024 * 1024;
+
   IndexGraph() = default;
   explicit IndexGraph(unique_ptr<GeometryLoader> loader, shared_ptr<EdgeEstimator> estimator);
 
   // Creates edge for points in same feature.
   void GetDirectedEdge(uint32_t featureId, uint32_t pointFrom, uint32_t pointTo, Joint::Id target,
                        bool forward, vector<JointEdge> & edges);
-  void GetNeighboringEdges(RoadPoint const & rp, bool isOutgoing, vector<JointEdge> & edges);
+  void GetNeighboringEdges(RoadPoint const & rp, bool isOutgoing, bool graphWithoutRestrictions,
+                           vector<JointEdge> & edges);
+  /// \brief Fills |edges| with edges going through |rp|. |rp| should not be a joint.
+  void GetIntermediatePointEdges(RoadPoint const & rp, bool graphWithoutRestrictions,
+                                 vector<DirectedEdge> & edges);
 
-  // Put outgoing (or ingoing) egdes for jointId to the 'edges' vector.
-  void GetEdgeList(Joint::Id jointId, bool isOutgoing, vector<JointEdge> & edges);
+  // Put outgoing (or ingoing) egdes for jointId to the |edges| vector.
+  void GetEdgeList(Joint::Id jointId, bool isOutgoing, bool graphWithoutRestrictions,
+                   vector<JointEdge> & edges);
+  void GetEdgeList(Joint::Id jointId, bool isOutgoing, bool graphWithoutRestrictions,
+                   vector<DirectedEdge> & edges);
   Joint::Id GetJointId(RoadPoint const & rp) const { return m_roadIndex.GetJointId(rp); }
   m2::PointD const & GetPoint(Joint::Id jointId);
 
-  Geometry & GetGeometry() { return m_geometry; }
   EdgeEstimator const & GetEstimator() const { return *m_estimator; }
   RoadJointIds const & GetRoad(uint32_t featureId) const { return m_roadIndex.GetRoad(featureId); }
 
   uint32_t GetNumRoads() const { return m_roadIndex.GetSize(); }
   uint32_t GetNumJoints() const { return m_jointIndex.GetNumJoints(); }
-  uint32_t GetNumPoints() const { return m_jointIndex.GetNumPoints(); }
+  uint32_t GetNumStaticPoints() const { return m_jointIndex.GetNumStaticPoints(); }
```

Comment:
GetNumStaticPoints returns number static RoadPoints. That mean that it's a number of RoadPoint which IndexGraph has before adding restrictions. 

If you're interested in dynamic road point please have a look at |JointIndex::m_dynamicJoints|.  m_dynamicJoints is a map from Joint:Id to Joint (with road points.)